File: tic-tac-toe-javafx-java/src/main/java/com/mj/tic/tac/toe/javafx/java/engine/MoveStrategy.java
package com.mj.tic.tac.toe.javafx.java.engine;

import com.mj.tic.tac.toe.javafx.java.util.Board;
import com.mj.tic.tac.toe.javafx.java.util.Coordinates;
import com.mj.tic.tac.toe.javafx.java.util.Player;
import java.util.Optional;

/**
 * Strategy interface for selecting a move in a tic-tac-toe game.
 * <p>
 * Implementations encapsulate different AI decision algorithms,
 * ranging from random selection to minimax-based optimal play.
 * Strategies are expected to be stateless or internally stateful
 * but should be safe for single-threaded use.
 *
 * <h2>Contract</h2>
 * <ul>
 *   <li>Implementations must <strong>not</strong> modify the board.</li>
 *   <li>If no move is available (board is full), the method must
 *       return {@link Optional#empty()}.</li>
 *   <li>The returned {@link Coordinates} (if present) must correspond
 *       to a cell that is empty at the time of the call.</li>
 * </ul>
 *
 * @author Montaser Jamal
 * @version 1.0
 * @email montaser.jjs@gmail.com
 * @phone +962-786258874
 * @see RandomMoveStrategy
 * @see Coordinates
 * @see Board
 * @since 30-05-2026
 */
public interface MoveStrategy {

    /**
     * Computes a move for the given player on the specified board.
     *
     * @param board  the current game state; must not be {@code null}.
     * @param player the player who is about to move; must not be
     *               {@code null}.
     * @return an {@link Optional} containing the chosen
     * {@link Coordinates}, or {@link Optional#empty()} if no
     * empty cells remain.
     */
    Optional<Coordinates> findMove(Board board, Player player);
}

File: tic-tac-toe-javafx-kotlin/src/main/kotlin/com/mj/tic/tac/toe/javafx/kotlin/engine/MoveStrategy.kt
package com.mj.tic.tac.toe.javafx.kotlin.engine

import com.mj.tic.tac.toe.javafx.kotlin.util.Board
import com.mj.tic.tac.toe.javafx.kotlin.util.Coordinates
import com.mj.tic.tac.toe.javafx.kotlin.util.Player
import java.util.Optional

/**
 * Functional interface for AI move-generation algorithms.
 *
 * Implementations encapsulate different strategies for choosing the
 * computer's next move on a Tic Tac Toe board. This interface is
 * defined as a Kotlin [fun interface] (SAM), allowing both class-based
 * implementations and lambda-based usage.
 *
 * The Strategy pattern is used here to decouple the AI move selection
 * logic from the task infrastructure, making it easy to add new
 * difficulty levels or move strategies without modifying existing code.
 *
 * @see RandomMoveStrategy Picks a random empty cell (used for [Difficulty.EASY]).
 * @see MinimaxMoveStrategy Uses minimax with alpha-beta pruning (used for [Difficulty.HARD]).
 * @author Montaser Sbaih
 * @version 1.0
 * @email montaser.jjs@gmail.com
 * @phone +962-786258874
 * @since 29-05-2026
 */

fun interface MoveStrategy {

    /**
     * Computes the best move for the given player on the current board.
     *
     * @param board The current game board state to analyze.
     * @param player The AI player making the move (typically [Player.COMPUTER]).
     * @return An [Optional] containing the chosen [Coordinates], or
     *   [Optional.empty] if no valid moves are available (board is full).
     */
    fun findMove(board: Board, player: Player): Optional<Coordinates>
}
